### PR TITLE
Test Cases for URL Generation

### LINKED ISSUE #109: URL Generation Does Not Take Forwarded Header Into Account
Running PREvant behind HTTPS breaking the URLs in the frontend because the [`RequestInfo` construction](https://github.com/aixigo/PREvant/blob/bc93f9ccce110b79a4de24367106a144a9e37bc0/api/src/models/request_info.rs) does not handle this case.

### PR DESCRIPTION
**Pull request by AnumDastgir**
Added testcases to ensure a successful URL generation after adding the code to deal with x-forwarded-header values.
Fixes #109

### CODE REVIEW

File: api/src/models/request_info.rs [hunk 1]
```diff
@@ -67,16 +45,32 @@ impl<'r> FromRequest<'r> for RequestInfo {
     type Error = ();
 
     async fn from_request(request: &'r Request<'_>) -> request::Outcome<Self, Self::Error> {
-        let hosts: Vec<_> = request.headers().get("host").collect();
+        let forwarded_host = request.headers().get_one("x-forwarded-host");
```
Comment (schrieveslaach) [inline api/src/models/request_info.rs:48]: This part of the code has been already merged in your previous PR #137. I think you have to rebase…

File: api/src/apps/routes.rs [hunk 2]
```diff
@@ -513,4 +513,522 @@ mod tests {
...
+                .mount("/", routes![crate::apps::routes::apps]);
+            let client = Client::tracked(rocket).await.expect("valid rocket");
+            let mut get = client.get(rocket::uri!(crate::apps::routes::apps));
+            get.add_header(ContentType::JSON);
+            let response = get.dispatch().await;
+            assert_eq!(response.status(), Status::BadRequest);
+        }
+
+        #[tokio::test]
+        async fn with_invalid_headers() {
```
Comment (schrieveslaach) [inline api/src/apps/routes.rs:853]: `with_invalid_headers`, `with_invalid_value_space_as_host`, and `with_invalid_value_backslash_as_host` test all the same equivalence class and thus more then one variant is sufficient.

A bit dated, but [this video](https://www.youtube.com/watch?v=P1Hv2sUPKeM) explains equivalence partitioning and what I mean by equivalence class.

File: api/src/apps/routes.rs [hunk 3]
```diff
@@ -513,4 +513,522 @@ mod tests {
...
+            let get = client
+                .get(rocket::uri!(crate::apps::routes::apps))
+                .header(Header::new("x-forwarded-proto", "."));
+
+            let response = get.dispatch().await;
+            assert_eq!(response.status(), Status::BadRequest);
+        }
+
+        #[tokio::test]
+        async fn with_valid_values() {
```
Comment (schrieveslaach) [inline api/src/apps/routes.rs:929]: This test tests nothing new, no new equivalence class… So the tests at the beginning should have you covered.

File: api/src/apps/routes.rs [hunk 4]
```diff
@@ -513,4 +513,522 @@ mod tests {
...
+                ]
+            });
+
+            assert_json_include!(actual: json, expected: expected_response);
+
+            Ok(())
+        }
+
+        #[tokio::test]
+        async fn without_prefer_header() {
```
Comment (schrieveslaach) [inline api/src/apps/routes.rs:836]: Typo or copy paste error?

```suggestion
        async fn bad_request_without_host_header() {
```

File: api/src/apps/routes.rs [hunk 5]
```diff
@@ -513,4 +513,522 @@ mod tests {
...
+            let client = Client::tracked(rocket).await.expect("valid rocket");
+            let get = client
+                .get(rocket::uri!(crate::apps::routes::apps))
+                .header(Header::new("x-forwarded-host", "localhost"));
+            let response = get.dispatch().await;
+            assert_eq!(response.status(), Status::Ok);
+        }
+
+        #[tokio::test]
+        async fn without_prefer_header2() {
```
Comment (schrieveslaach) [inline api/src/apps/routes.rs:947]: Seems to be the same test as above. Please, remove.

File: api/src/apps/routes.rs [hunk 6]
```diff
@@ -513,4 +513,522 @@ mod tests {
...
+                    None,
+                    &vec![sc!("service-a")],
+                )
+                .await?;
+            host_meta_crawler.fake_host_meta_info("master".to_string(), "service-a".to_string());
+            let rocket = rocket::build()
+                .manage(host_meta_cache)
+                .manage(apps)
+                .mount("/", routes![crate::apps::routes::apps])
+                .mount("/api/apps", crate::apps::apps_routes());
```
Comment (schrieveslaach) [inline api/src/apps/routes.rs:552]: This block appears to be the very same in a lot of tests. You could create a function for it.

```suggestion
            let rocket = set_up_rochet_with_dummy_infrastructure_and_a_running_app();
```

File: api/src/apps/routes.rs [hunk 7]
```diff
@@ -513,4 +513,522 @@ mod tests {
...
+                ]
+            });
+
+            assert_json_include!(actual: json, expected: expected_response);
+
+            Ok(())
+        }
+
+        #[tokio::test]
+        async fn host_header_response_with_xforwardedproto_and_other_default_values(
```
Comment (schrieveslaach) [inline api/src/apps/routes.rs:598]: apply suggestions from `host_header_response_with_xforwardedhost_xforwardedproto_and_xforwardedport`

File: api/src/apps/routes.rs [hunk 8]
```diff
@@ -513,4 +513,522 @@ mod tests {
...
+                ]
+            });
+
+            assert_json_include!(actual: json, expected: expected_response);
+
+            Ok(())
+        }
+
+        #[tokio::test]
+        async fn host_header_response_with_xforwardedhost_and_other_default_values(
```
Comment (schrieveslaach) [inline api/src/apps/routes.rs:657]: apply suggestions from `host_header_response_with_xforwardedhost_xforwardedproto_and_xforwardedport`

File: api/src/apps/routes.rs [hunk 9]
```diff
@@ -513,4 +513,522 @@ mod tests {
...
+                ]
+            });
+
+            assert_json_include!(actual: json, expected: expected_response);
+
+            Ok(())
+        }
+
+        #[tokio::test]
+        async fn host_header_response_with_xforwardedport_and_default_values(
```
Comment (schrieveslaach) [inline api/src/apps/routes.rs:715]: apply suggestions from `host_header_response_with_xforwardedhost_xforwardedproto_and_xforwardedport`

File: api/src/apps/routes.rs [hunk 10]
```diff
@@ -513,4 +513,522 @@ mod tests {
...
+                ]
+            });
+
+            assert_json_include!(actual: json, expected: expected_response);
+
+            Ok(())
+        }
+
+        #[tokio::test]
+        async fn host_header_response_with_all_default_values(
```
Comment (schrieveslaach) [inline api/src/apps/routes.rs:778]: apply suggestions from `host_header_response_with_xforwardedhost_xforwardedproto_and_xforwardedport`

**Review by schrieveslaach — CHANGES_REQUESTED** [review]
Thanks for creating these test and I'm very pleased to see them.

I created a set of suggestion comments to your PR. Please, check and apply them.

**Comment by AnumDastgir** [comment]
Thank you so much for such detailed suggestions. I have implemented and committed them.

**Review by schrieveslaach — APPROVED** [review]
Changes look good to me. Please, squash your commits into one and then I'll merge it to main.

**Pull request merged**